app/segments/drill/[id]/submission/input.js

```diff
@@ -137,6 +141,136 @@ async function uploadAttempt(
 }
 
 //TODO: Create a function to check leaderboard and update if needed
+function checkLeaderboardUpdate(uploadData, drillInfo, currentLeaderboard) {
+  const leaderboardData = currentLeaderboard["data"];
+
+  //used if an attempt already exists
+  const currentBest =
+    leaderboardData[uploadData.uid]?.[drillInfo.mainOutputAttempt]?.value;
+
+  const newAttempt = {
+    [drillInfo.mainOutputAttempt]: {
+      id: uploadData.id,
+      value: uploadData[drillInfo.mainOutputAttempt],
+    },
+  };
+
+  console.log("currentBest: ", currentBest);
+
+  //check if the user exists on the leaderboard
+  if (leaderboardData[uploadData.uid] == undefined) {
+    console.log("User not on leaderboard, uploading this attempt");
+
+    uploadNewLeaderboard(
+      leaderboardData,
+      newAttempt,
+      uploadData.uid,
+      uploadData.did,
+    );
+  } else {
+    //Determine if lower is better for mainOutputAttempt
+    if (drillInfo.aggOutputs[drillInfo.mainOutputAttempt].lowerIsBetter) {
+      console.log("Lower is in fact better");
+      //compare the current attempt to the attempt currently on leaderboard
+      if (uploadData[drillInfo.mainOutputAttempt] < currentBest) {
+        console.log("New Best Attempt! Time to upload!");
+
+        uploadNewLeaderboard(
+          leaderboardData,
+          newAttempt,
+          uploadData.uid,
+          uploadData.did,
+        );
+      }
+      //else for testing
+      else {
+        console.log("Didn't update");
+      }
+    } else {
+      // console.log(uploadData)
+      console.log("Lower is not better");
+      //compare the current attempt to the attempt currently on leaderboard
+      if (uploadData[drillInfo.mainOutputAttempt] > currentBest) {
+        console.log("New Best Attempt! Time to upload!");
+
+        uploadNewLeaderboard(
+          leaderboardData,
+          newAttempt,
+          uploadData.uid,
+          uploadData.did,
+        );
+      }
+      //else for testing
+      else {
+        console.log("Didn't update");
+      }
+    }
+  }
+}
+
+async function uploadNewLeaderboard(leaderboardData, newAttempt, uid, did) {
+  //Add new attempt to leaderboard
+  leaderboardData[uid] = newAttempt;
+  const submitMainOutputAttempt = Object.keys(newAttempt)[0];
+  console.log("new attempt");
+  console.log(newAttempt);
+
+  //Reference to best_attempts drill document
+  const bestAttemptsDrillRef = doc(db, "teams", "1", "best_attempts", did);
+
+  try {
+    console.log("LEADERBOARD UPDATE STARTED");
+    console.log(submitMainOutputAttempt);
+
+    // get a fresh call without tanstack cache just in case
+    const latestLeaderboard = await getDoc(bestAttemptsDrillRef);
```

Comment:
If this is done anyways do we need to pass leaderboardData?

Reply:
I think I was just concerned the data might have gotten out of date / stale cache in the meantime since the `checkLeaderboardUpdate` function, which was why I called `getDoc` again.

I saw the other comment suggesting to just initialize newAttempt in updateNewLeaderboard function, I will try that and/or the firebase transactions thing to avoid race case etc.

Reply:
stopped passing in leaderboardData